pandas_gbq/gbq.py

```diff
@@ -566,6 +570,104 @@ def run_query(self, query, **kwargs):
 
         return schema, result_rows
 
+    def export_table_to_gcs(
+        self,
+        dataset,
+        table,
+        timeout_in_seconds=None,
+        bucket=None,
+        blob=None,
+        zipped=True,
+    ):
+        """
+        export table to gcs.  returns tuple of (bucket, blob)
+        """
+        client = self.client
+        table_ref = client.dataset(dataset).table(table)
+        job_config = ExtractJobConfig()
+        job_config.compression = 'GZIP' if zipped else 'NONE'
+        bucket = bucket or '{}-temp'.format(client.project)
+        blob = blob or '{}/{}.csv'.format(dataset, table)
+        if zipped and not blob.endswith('.gz'):
+            blob += '.gz'
+        if not isinstance(bucket, str):
+            bucket = bucket.name
+        destination_uri = 'gs://{}/{}'.format(bucket, blob)
+        extract_job = client.extract_table(
+            table_ref, destination_uri, job_config=job_config)
+        wait_for_job(extract_job, timeout_in_seconds=timeout_in_seconds)
+        logger.info('Exported {}.{} -> {}'.format(
+            dataset,
+            table,
+            destination_uri)
+        )
+        return bucket, blob
+
+    def read_gbq_table(
+        self,
+        dataset,
+        table,
+        bucket,
+        timeout_in_seconds=600,
+    ):
+        """
+        reads an entire table from gbq into a dataframe
+        """
+        from google.cloud import storage
+
+        storage = storage.Client(self.project, self.credentials)
+        prefix = 'gbq-exports/{}/{}/'.format(dataset, table)
+        bucket = storage.get_bucket(bucket)
+
+        for old_blob in bucket.list_blobs(prefix=prefix):
+            old_blob.delete()
+            logger.info('Old Blob Deleted: {}'.format(old_blob.name))
+
+        self.export_table_to_gcs(
+            dataset=dataset,
+            table=table,
+            timeout_in_seconds=timeout_in_seconds,
+            bucket=bucket,
+            blob='{}*.csv.gz'.format(prefix),
+        )
+
+        frames = []
+
+        downloads = tqdm(list(bucket.list_blobs(prefix=prefix)), unit='file')
+        for blob in downloads:
+            downloads.set_description('Processing {}, {}MB'.format(
+                blob, blob.size / 2**20))
+            s = BytesIO(blob.download_as_string())
+            frames.append(pd.read_csv(s, compression='gzip'))
+            blob.delete()
+
+        return pd.concat(frames, ignore_index=True)
+
+    def read_gbq_bulk(
```

Comment:
There's going to be some threshold for query results size where it's going to always make sense to do this (not sure exactly what that threshold is right now). For we could look at the number of rows times the number of columns in the query results to figure out which method to use.

Though, since this requires a staging bucket and regular queries don't, you're probably right that it should be explicitly set. I'd like to see us refactor the `read_gbq` method a bit so that we could reuse the auth & query running portion without the query downloading portion.

One option might be: add a `storage_bucket` argument to `read_gbq`. If set, use the bulk download method, otherwise use current method.

Reply:
Yes, agree on the tradeoffs. I think we should at least start by being explicit (maybe a user warning for big queries?), and could then think about defaulting

I think the cleavage is around 500k rows